File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/ir/UnconditionalJumpStatement.java
package ro.pub.cs.diploma.ir;

import com.intellij.openapi.util.Ref;
import org.jetbrains.annotations.NotNull;

public class UnconditionalJumpStatement implements JumpStatement {
  @NotNull private final Ref<Block> blockRef;

  public UnconditionalJumpStatement(@NotNull Ref<Block> blockRef) {
    this.blockRef = blockRef;
  }

  @NotNull
  public Block getBlock() {
    return blockRef.get();
  }

  @Override
  public void accept(@NotNull final Visitor visitor) {
    visitor.visit(this);
  }

  @Override
  public void detach() {
    blockRef.get().removeReference(blockRef);
  }
}

File: inspection/src/ro/pub/cs/diploma/ir/UnconditionalJumpStatement.kt
package ro.pub.cs.diploma.ir

import com.intellij.openapi.util.Ref

class UnconditionalJumpStatement(val blockRef: Ref<Block>) : TerminatorStatement {
  val block: Block
    get() = blockRef.get()

  override fun accept(visitor: Visitor) {
    visitor.visit(this)
  }
}
